### PR TITLE
CXD-16 Enhance the Asset Import package for Data Products

### PR DESCRIPTION
**Pull request by ErnestoLoma**

### CODE REVIEW

File: package-toolkit/runtime/src/main/kotlin/com/atlan/pkg/cache/DataDomainCache.kt [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                addByGuid(dataDomain.guid, dataDomain)
+            }
+        DataDomain.select()
+            .includesOnResults(includesOnResults)
+            .where(DataDomain.PARENT_DOMAIN_QUALIFIED_NAME.hasAnyValue())
+            .sort(DataDomain.PARENT_DOMAIN_QUALIFIED_NAME.order(SortOrder.Asc))
+            .stream(false)
+            .forEach { dataDomain ->
+                addByGuid(dataDomain.guid, dataDomain)
+            }
```
Comment (cmgrote) [inline package-toolkit/runtime/src/main/kotlin/com/atlan/pkg/cache/DataDomainCache.kt:92]: Just curious — why parallel-stream the first search and sequentially-stream the second?

File: package-toolkit/runtime/src/main/kotlin/com/atlan/pkg/serde/cell/DataDomainXformer.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                } else {
+                    ""
+                }
+            }
+            else -> AssetRefXformer.encode(asset)
+        }
+    }
+
+    /**
+     * Decodes (deserializes) a string form into a category reference object.
```
Comment (cmgrote) [inline package-toolkit/runtime/src/main/kotlin/com/atlan/pkg/serde/cell/DataDomainXformer.kt:38]: Copy / paste here refers to `category reference` should be `data domain reference` (maybe other places, too?)

File: samples/packages/asset-import/src/main/kotlin/com/atlan/pkg/aim/DomainImporter.kt [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    updateOnly = updateOnly,
+    batchSize = batchSize,
+    trackBatches = true, // Always track batches for GTC importers, to ensure cache is managed
+    fieldSeparator = fieldSeparator,
+) {
+    private var levelToProcess = 0
+
+    // Maximum depth of any domain in the CSV -- will be updated on first pass through the CSV
+    // file by includeRow() method
+    private val maxCategoryDepth = AtomicInteger(1)
```
Comment (cmgrote) [inline samples/packages/asset-import/src/main/kotlin/com/atlan/pkg/aim/DomainImporter.kt:53]: Change variables / comments regarding `category` to `domain`

File: samples/packages/asset-import/src/main/resources/package.pkl [hunk 4]
```diff
@@ -297,5 +397,26 @@ uiConfig {
         "glossaries_batch_size"
       }
     }
+    new UIRule {
+          whenInputs { ["data_products_import_type"] = "UPLOAD" }
+          required { "data_products_file" }
+        }
```
Comment (cmgrote) [inline samples/packages/asset-import/src/main/resources/package.pkl:403]: fix spacing — 2 spaces per indent level

File: samples/packages/asset-import/src/main/resources/package.pkl [hunk 5]
```diff
@@ -251,6 +251,106 @@ uiConfig {
         }
       }
     }
+    ["Data Products"] {
```
Comment (cmgrote) [inline samples/packages/asset-import/src/main/resources/package.pkl:254]: fix spacing — 2 spaces per indent level

File: samples/packages/asset-import/src/main/resources/package.pkl [hunk 6]
```diff
@@ -251,6 +251,106 @@ uiConfig {
...
+              title = "S3 region"
+              required = false
+              helpText = "Enter the S3 region from which to retrieve the S3 object. If empty, will use the region of Atlan's own back-end storage."
+              placeholderText = "ap-south-1"
+              width = 4
+            }
+            ["data_products_s3_bucket"] = new TextInput {
+              title = "S3 bucket"
+              required = false
+              helpText = "Enter the S3 bucket from which to retrieve the S3 object. IF empty, will use the bucket of Atlan's own back-end storage."
```
Comment (cmgrote) [inline samples/packages/asset-import/src/main/resources/package.pkl:284]: `IF` -> `If` (maybe a typo elsewhere in here, too?)

File: samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+    @BeforeClass
+    fun beforeClass() {
+        prepFile()
+//        createTags()
+        setup(
+            AssetImportCfg(
+                assetsFile = null,
+                assetsUpsertSemantic = null,
+                assetsAttrToOverwrite = null,
```
Comment (cmgrote) [inline samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt:112]: Any config options you don't specify should already be set to `null` by default, so might be easiest to just leave them out entirely here.

File: samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        assertEquals(CertificateStatus.VERIFIED, d1.certificateStatus)
+        val d2 = level2_domains[0]
+        assertEquals(dataDomain2, d2.name)
+        assertEquals("Test subdomain", d2.userDescription)
+        assertEquals(setOf("ernest"), d2.ownerUsers)
+        assertEquals(setOf("admins"), d2.ownerGroups)
+        assertEquals(d1.guid, d2.parentDomain.guid)
+        assertEquals(d2.qualifiedName, DataDomain.generateQualifiedName(IDataMesh.generateSlugForName(dataDomain2), d1.qualifiedName))
+        assertEquals(d2.parentDomainQualifiedName, d1.qualifiedName)
+        assertEquals(d1.qualifiedName, d2.superDomainQualifiedName)
```
Comment (cmgrote) [inline samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt]: Should be in a separate test function (e.g. `domain2Created()`)

File: samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        assertEquals("With a message!", d2.certificateStatusMessage)
+        val d3 = level3_domains[0]
+        assertEquals(dataDomain3, d3.name)
+        assertEquals("Test sub sub domain", d3.userDescription)
+        assertEquals(setOf("ernest"), d3.ownerUsers)
+        assertEquals(setOf("admins"), d3.ownerGroups)
+        assertEquals(d2.guid, d3.parentDomain.guid)
+        assertEquals(d3.qualifiedName, DataDomain.generateQualifiedName(IDataMesh.generateSlugForName(dataDomain3), d2.qualifiedName))
+        assertEquals(d3.parentDomainQualifiedName, d2.qualifiedName)
+        assertEquals(d1.qualifiedName, d3.superDomainQualifiedName)
```
Comment (cmgrote) [inline samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt]: Separate test function

File: samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        val p1 = findDataProductWithRetry(dataProduct1)
+        assertEquals(dataProduct1, p1.name)
+        assertEquals("Test data product", p1.userDescription)
+        assertEquals(setOf("ernest"), p1.ownerUsers)
+        assertEquals(setOf("admins"), p1.ownerGroups)
+        assertEquals(d3.qualifiedName, p1.dataDomain.qualifiedName)
```
Comment (cmgrote) [inline samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt:172]: Separate test function

File: samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportGlossariesTest.kt
Comment (cmgrote) [inline samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportGlossariesTest.kt:1]: This doesn't seem to fit (totally unrelated to any data products or domains work)... Did you intend to commit this file?

File: sdk/src/main/java/com/atlan/model/assets/DataProduct.java
Comment (cmgrote) [inline sdk/src/main/java/com/atlan/model/assets/DataProduct.java:1]: Need to add this method to the Freemarker template used by the generator: `sdk/src/main/resources/templates/DataProduct.ftl`

**Review by cmgrote — CHANGES_REQUESTED** [review]

File: package-toolkit/runtime/src/main/kotlin/com/atlan/pkg/cache/DataDomainCache.kt (hunk 1 above)
Comment (ErnestoLoma) [reply package-toolkit/runtime/src/main/kotlin/com/atlan/pkg/cache/DataDomainCache.kt:92]: That wasn't intentional. Not sure how it happened. I have changed it.

File: package-toolkit/runtime/src/main/kotlin/com/atlan/pkg/serde/cell/DataDomainXformer.kt (hunk 2 above)
Comment (ErnestoLoma) [reply package-toolkit/runtime/src/main/kotlin/com/atlan/pkg/serde/cell/DataDomainXformer.kt:38]: fixed.

File: samples/packages/asset-import/src/main/kotlin/com/atlan/pkg/aim/DomainImporter.kt (hunk 3 above)
Comment (ErnestoLoma) [reply samples/packages/asset-import/src/main/kotlin/com/atlan/pkg/aim/DomainImporter.kt:53]: done

File: samples/packages/asset-import/src/main/resources/package.pkl (hunk 4 above)
Comment (ErnestoLoma) [reply samples/packages/asset-import/src/main/resources/package.pkl:403]: Fixed

File: sdk/src/main/java/com/atlan/model/assets/DataProduct.java
Comment (ErnestoLoma) [reply sdk/src/main/java/com/atlan/model/assets/DataProduct.java:1]: Done

File: samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt (hunk 9 above)
Comment (ErnestoLoma) [reply samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt]: Done

File: samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt (hunk 10 above)
Comment (ErnestoLoma) [reply samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt:186]: Done

File: samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt (hunk 7 above)
Comment (ErnestoLoma) [reply samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportDataDomainTest.kt]: Changed.

File: samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportGlossariesTest.kt
Comment (ErnestoLoma) [reply samples/packages/asset-import/src/test/kotlin/com/atlan/pkg/aim/ImportGlossariesTest.kt:1]: Removed

File: samples/packages/asset-import/src/main/resources/package.pkl (hunk 6 above)
Comment (ErnestoLoma) [reply samples/packages/asset-import/src/main/resources/package.pkl]: Fixed

**Review by cmgrote — APPROVED** [review]

**Pull request merged**